src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java

```diff
@@ -0,0 +1,170 @@
+
+import org.apache.commons.lang3.ArrayUtils;
+
+import net.minecraft.command.CommandException;
+import net.minecraft.command.EntitySelector;
+import net.minecraft.command.ICommandSender;
+import net.minecraft.entity.Entity;
+import net.minecraftforge.fml.common.FMLLog;
+import net.minecraftforge.fml.common.Loader;
+
+/**
+ * Allows registration of custom selector types by assigning a {@link SelectorHandler} to a prefix
+ * To check whether a new prefix conflicts with any registered prefix, the methods {@link #getShadowed(String)} and {@link #getShadowing(String)} can be used
+ * This class handles calls to the {@link EntitySelector} methods {@link EntitySelector#matchEntities matchEntities},
+ * {@link EntitySelector#matchesMultiplePlayers matchesMultiplePlayers} and {@link EntitySelector#hasArguments hasArguments}.<br>
+ * The calls are delegated to the handler with the longest matching prefix
+ */
+public class SelectorHandlerManager
+{
+    private SelectorHandlerManager()
+    {
+    }
+
+    //the ordering is reversed such that longer prefixes appear before their shorter substrings
+    private static final NavigableMap<String, SelectorHandler> selectorHandlers = new TreeMap<String, SelectorHandler>(Collections.<String> reverseOrder());
+    private static final NavigableMap<String, String> registeringMods = new TreeMap<String, String>(Collections.<String> reverseOrder());
+
+    private static final SelectorHandler vanillaHandler = new SelectorHandler()
+    {
+        @Override
+        public <T extends Entity> List<T> matchEntities(final ICommandSender sender, final String token, final Class<? extends T> targetClass) throws CommandException
+        {
+            return EntitySelector.matchEntitiesDefault(sender, token, targetClass);
+        }
+
+        @Override
+        public boolean matchesMultiplePlayers(final String selectorStr) throws CommandException
+        {
+            return EntitySelector.matchesMultiplePlayersDefault(selectorStr);
+        }
+
+        @Override
+        public boolean isSelector(final String selectorStr)
+        {
+            return EntitySelector.isSelectorDefault(selectorStr);
+        }
+    };
+
+    static
+    {
+        for (final String prefix : ArrayUtils.toArray("@p", "@a", "@r", "@e"))
+        {
+            selectorHandlers.put(prefix, vanillaHandler);
+            registeringMods.put(prefix, "minecraft");
+        }
+    }
+
+    /**
+     * Returns a collection of all registered prefixes more specific than {@code prefix}
+     *
+     * @return A {@link Map} mapping the prefixes to the mod ids registering them
+     */
+    public static Map<String, String> getShadowing(final String prefix)
+    {
+        Map<String, String> ret = new HashMap<String, String>();
+
+        for (final Entry<String, String> other : registeringMods.headMap(prefix).entrySet())
+            if (other.getKey().startsWith(prefix))
+                ret.put(other.getKey(), other.getValue());
+
```

Comment:
The point is, if you want to find all registered prefixes starting with "a", it makes sense to iterate through the map starting at "a". As it stands, using `headMap`, "a" would be the last element visited. Hence the `descendingMap` call.

Then, when looking for prefixes starting with "a", you can stop when you get to "b", because no subsequent prefixes will satisfy the condition.

Doing so should still return all colliding prefixes, the aim is simply to avoid unnecessarily checking elements.